src/xmerl_c14n.erl

```diff
@@ -0,0 +1,527 @@
+    NsPartRaw = case Ns of
+        empty -> Nsp#xmlNamespace.default;
+        [] ->
+            if Nsp == [] -> 'urn:oasis:names:tc:SAML:2.0:assertion';
+               true -> Nsp#xmlNamespace.default
+            end;
+        _ ->
+            case proplists:get_value(Ns, Nsp#xmlNamespace.nodes) of
+                undefined ->
+                    error({ns_not_found, Ns, Nsp});
+                Uri -> atom_to_list(Uri)
+            end
+    end,
+    NsPart = if is_atom(NsPartRaw) -> atom_to_list(NsPartRaw); true -> NsPartRaw end,
+    NamePart = if is_atom(Name) -> atom_to_list(Name); true -> Name end,
+    lists:flatten([NsPart | NamePart]).
+
+%% @doc Returns the canonical URI name of an XML element or attribute.
+%% @private
+-spec canon_name(#xmlElement{} | #xmlAttribute{}) -> string().
+canon_name(#xmlAttribute{name = Name, nsinfo = Exp, namespace = Nsp}) ->
+    case Exp of
+        {Ns, Nme} -> canon_name(Ns, Nme, Nsp);
+        _ -> canon_name([], Name, Nsp)
+    end;
+canon_name(#xmlElement{name = Name, nsinfo = Exp, namespace = Nsp}) ->
+    case Exp of
+        {Ns, Nme} -> canon_name(Ns, Nme, Nsp);
+        _ -> canon_name([], Name, Nsp)
+    end.
+
+%% @doc Compares two XML attributes for c14n purposes
+-spec attr_lte(A :: #xmlAttribute{}, B :: #xmlAttribute{}) -> true | false.
+attr_lte(AttrA, AttrB) ->
+    logger:debug("AttrA: ~p AttrB: ~p~n", [AttrA, AttrB]),
+    A = canon_name(AttrA), B = canon_name(AttrB),
+    PrefixedA = case AttrA#xmlAttribute.nsinfo of {_, _} -> true; _ -> false end,
+    PrefixedB = case AttrB#xmlAttribute.nsinfo of {_, _} -> true; _ -> false end,
+    if (PrefixedA) andalso (not PrefixedB) ->
+        false;
+    (not PrefixedA) andalso (PrefixedB) ->
+        true;
+    true ->
+        A =< B
+    end.
+
+%% @doc Cleans out all namespace definitions from an attribute list and returns it sorted.
+%% @private
+-spec clean_sort_attrs(Attrs :: [#xmlAttribute{}]) -> [#xmlAttribute{}].
+clean_sort_attrs(Attrs) ->
+    lists:sort(fun(A,B) ->
+        attr_lte(A, B)
+    end, lists:filter(fun(Attr) ->
+        case Attr#xmlAttribute.nsinfo of
+            {"xmlns", _} -> false;
+            _ -> case Attr#xmlAttribute.name of
+                'xmlns' -> false;
+                _ -> true
+            end
+        end
+    end, Attrs)).
+
+%% @doc Returns the list of namespace prefixes "needed" by an element in canonical form
+%% @private
+-spec needed_ns(Elem :: #xmlElement{}, InclNs :: [string()]) -> [string()].
+needed_ns(#xmlElement{nsinfo = NsInfo, attributes = Attrs}, InclNs) ->
+    NeededNs1 = case NsInfo of
+        {Nas, _} -> [Nas];
+        _ -> []
+    end,
+    %% show through namespaces that apply at the bottom level? this part of the spec is retarded
+    %%KidElems = [K || K <- Kids, element(1, K) =:= xmlElement],
+    NeededNs2 = NeededNs1, %case KidElems of
+        %%[] -> [K || {K,V} <- E#xmlElement.namespace#xmlNamespace.nodes];
+        %%_ -> NeededNs1
+    %%end,
```

Comment:
Remove the commented out code, and this bind seems useless.

Reply:
That was in the original code. Wanted to preserve it cause it shows some intent of the original author.